Build failure: golly [darwin]

### Nixpkgs version

- Unstable (25.05)

### Steps to reproduce

```
$ NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1 nix build --impure github:NixOS/nixpkgs/1362f7bd36ea54444eab7a88294677e6660fd25a#legacyPackages.x86_64-darwin.golly
```

### Can Hydra reproduce this build failure?

Please select the Hydra Status.

### Link to Hydra build job

_No response_

### Relevant log output

```console
error: builder for '/nix/store/f72zhj0kqkaqc9g5pfv7q0i613q4ddld-golly-4.3.drv' failed with exit code 2;
       last 14 log lines:
       > Running phase: unpackPhase
       > unpacking source archive /nix/store/xpixnx2a75vwvr607l3ihx28030f4anz-golly-4.3-src.tar.gz
       > source root is golly-4.3-src/gui-wx
       > setting SOURCE_DATE_EPOCH to timestamp 1716282223 of file "golly-4.3-src/gui-wx/makefile-win"
       > Running phase: patchPhase
       >     pythonlib = wxT(STRINGIFY(/nix/store/g13l84j8gn1fdhl3pzkg2r9ip9jz1hhd-python3-3.12.10/lib/libpython3.so));
       > Running phase: configurePhase
       > no configure script, doing nothing
       > Running phase: buildPhase
       > build flags: SHELL=/nix/store/1ys14sy4jzaclkfrwvnqvid0d8p0lawh-bash-5.2p37/bin/bash -f makefile-gtk ENABLE_SOUND=1 GOLLYDIR=/nix/store/y8k28p2d5bfcmm94hcbzvvs02km46slq-golly-4.3/share/golly
       > mkdir -p ObjGTK
       > g++ -DVERSION=4.3 -DGOLLYDIR="/nix/store/y8k28p2d5bfcmm94hcbzvvs02km46slq-golly-4.3/share/golly" -D_FILE_OFFSET_BITS=64 -D_LARGE_FILES -I../gollybase -O3 -Wall -Wno-non-virtual-dtor -fno-strict-aliasing  -DENABLE_SOUND -c -o ObjGTK/bigint.o ../gollybase/bigint.cpp
       > /nix/store/1ys14sy4jzaclkfrwvnqvid0d8p0lawh-bash-5.2p37/bin/bash: line 1: g++: command not found
       > make: *** [makefile-gtk:141: ObjGTK/bigint.o] Error 127
       For full logs, run:
         nix log /nix/store/f72zhj0kqkaqc9g5pfv7q0i613q4ddld-golly-4.3.drv
```

### Additional context

_No response_

### System metadata

 - system: `"aarch64-darwin"`
 - host os: `Darwin 24.5.0, macOS 15.5`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.28.3`
 - nixpkgs: `/nix/store/04bywqz2ys8xscpg7nmmw8ykf3fasgqc-source`



### Notify maintainers



---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/applications/science/misc/golly/default.nix
+++ b/pkgs/applications/science/misc/golly/default.nix
@@ -11,13 +11,13 @@
   libX11,
   SDL2,
 }:
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (finalAttrs: {
   pname = "golly";
   version = "4.3";
 
   src = fetchurl {
     hash = "sha256-UdJHgGPn7FDN4rYTgfPBAoYE5FGC43TP8OFBmYIqCB0=";
-    url = "mirror://sourceforge/project/golly/golly/golly-${version}/golly-${version}-src.tar.gz";
+    url = "mirror://sourceforge/project/golly/golly/golly-${finalAttrs.version}/golly-${finalAttrs.version}-src.tar.gz";
   };
 
   buildInputs = [
@@ -35,25 +35,40 @@ stdenv.mkDerivation rec {
     wrapGAppsHook3
   ];
 
+  # fails nondeterministically on darwin
+  enableParallelBuilding = false;
+
   setSourceRoot = ''
     sourceRoot=$(echo */gui-wx)
   '';
 
-  postPatch = ''
-    sed -e 's@PYTHON_SHLIB@${python3}/lib/libpython3.so@' -i wxprefs.cpp
-    ! grep _SHLIB *.cpp
-
-    grep /lib/libpython wxprefs.cpp
-  '';
+  postPatch =
+    ''
+      substituteInPlace wxprefs.cpp \
+        --replace-fail 'PYTHON_SHLIB' '${python3}/lib/libpython3.so'
+    ''
+    + lib.optionalString stdenv.hostPlatform.isDarwin ''
+      substituteInPlace makefile-gtk \
+        --replace-fail '-Wl,--as-needed' "" \
+        --replace-fail '-lGL ' "" \
+        --replace-fail '-lGLU' ""
+    '';
 
   makeFlags = [
     "-f"
     "makefile-gtk"
     "ENABLE_SOUND=1"
     "GOLLYDIR=${placeholder "out"}/share/golly"
+    "CC=${stdenv.cc.targetPrefix}cc"
+    "CXX=${stdenv.cc.targetPrefix}c++"
+    "CXXC=${stdenv.cc.targetPrefix}c++"
+    "LD=${stdenv.cc.targetPrefix}c++"
+    "WX_CONFIG=${lib.getExe' (lib.getDev wxGTK) "wx-config"}"
   ];
 
   installPhase = ''
+    runHook preInstall
+
     mkdir -p "$out/bin"
     cp ../golly ../bgolly "$out/bin"
 
@@ -62,13 +77,19 @@ stdenv.mkDerivation rec {
 
     mkdir -p "$out/share/golly"
     cp -r ../{Help,Patterns,Scripts,Rules} "$out/share/golly"
+
+    runHook postInstall
   '';
 
   meta = {
     description = "Cellular automata simulation program";
-    license = lib.licenses.gpl2;
-    maintainers = [ lib.maintainers.raskin ];
-    platforms = lib.platforms.linux;
+    license = lib.licenses.gpl2Plus;
+    maintainers = with lib.maintainers; [
+      raskin
+      siraben
+    ];
+    platforms = lib.platforms.unix;
+    homepage = "https://golly.sourceforge.io/";
     downloadPage = "https://sourceforge.net/projects/golly/files/golly";
   };
-}
+})

